### DISCUSSION TITLE
Perplessità per riprogrammare firmware esp32 con esphome [risolto]

### DISCUSSION THREAD

**Discussion by gualo1983** [indomus/forum#368] [Hardware & Firmware]
Buona sera,
sto cercando di realizzare il bridge/gateway bluetooth wi-fi come da [guida](https://indomus.it/progetti/realizzare-un-bridge-gateway-bluetooth-wi-fi-autonomo-con-nodemcu-esp32-e-esphome/).
Ho collegato il [NodeMCU](https://www.amazon.it/gp/product/B071P98VTG/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&th=1) al raspberry pi 3b+ ma quando provo a cercare a che porta logica è collegato non mi risulta nulla... 
ho provato ad usare lo script suggerito da [indomus](https://indomus.it/guide/utilita-elencare-le-interfacce-hardware-in-uso-su-raspbian-o-su-linux-piu-genericamente/)
ma il risultato è questo:

> pi@raspberrypi:~ $ ./ports.sh 
/dev/bus/usb/001/001 - Linux_5.10.17-v7+_dwc_otg_hcd_DWC_OTG_Controller_3f980000.usb
/dev/bus/usb/001/007 - dresden_elektronik_ingenieurtechnik_GmbH_ConBee_II_DE2152921
/dev/ttyACM0 - dresden_elektronik_ingenieurtechnik_GmbH_ConBee_II_DE2152921
/dev/bus/usb/001/003 - 0424_2514
/dev/sg0 - 
/dev/sda2 - SABRENT_SABRENT_DB9876543214E-0:0
/dev/sda - SABRENT_SABRENT_DB9876543214E-0:0
/dev/sda3 - SABRENT_SABRENT_DB9876543214E-0:0
/dev/sda1 - SABRENT_SABRENT_DB9876543214E-0:0
/dev/bsg/0:0:0:0 - 
/dev/bus/usb/001/005 - SABRENT_SABRENT_DB9876543214E
/dev/bus/usb/001/006 - 0424_7800
/dev/bus/usb/001/002 - 0424_2514

Sul Raspberry come si nota il SO è su USB e vi è collegata anche l'antenna ConBee_II
Purtroppo per continuare nella mia riprogrammazione devo capire come vedere la porta logica del nodemcu.

**Comment by snakuzzo** [comment]
devi tenere premuto il tastino FLASH sul nodemcu mentre lo inserisci nella porta USB. L'hai fatto ?

**Reply by xericcardo** [reply]
Ciao 
stesso problema anche collegandolo a windows non rileva la porta,anche dopo aver installato i driver.

**Reply by gualo1983** [reply]
> devi tenere premuto il tastino FLASH sul nodemcu mentre lo inserisci nella porta USB. L'hai fatto ?

no, non lo avevo fatto...
ho scoperto che pure il cavo usb-microusb che usavo non era adeguato. Appena lo ho sostituto e ho premuto il tastino all'inserimento della usb nel raspberry ha trovato il nodemcu.

Ora ho creato il file yaml e lo ho compilato con successo ma quando mi domanda se fare l'upload sul dispositivo (scelta 1) della procedura da riga di comando al momento di connettersi al nodemcu la procedura si blocca con questi errori 

> Uploading .pioenvs/blt2wifi01/firmware.bin
Serial port /dev/ttyUSB0
Connecting........_____
Traceback (most recent call last):
  File "/root/.platformio/packages/tool-esptoolpy/esptool.py", line 3969, in <module>
    _main()
  File "/root/.platformio/packages/tool-esptoolpy/esptool.py", line 3962, in _main
    main()
  File "/root/.platformio/packages/tool-esptoolpy/esptool.py", line 3552, in main
    esp.connect(args.before, args.connect_attempts)
  File "/root/.platformio/packages/tool-esptoolpy/esptool.py", line 519, in connect
    last_error = self._connect_attempt(mode=mode, esp32r0_delay=False)
  File "/root/.platformio/packages/tool-esptoolpy/esptool.py", line 477, in _connect_attempt
    self._setDTR(False)  # IO0=HIGH
  File "/root/.platformio/packages/tool-esptoolpy/esptool.py", line 443, in _setDTR self._port.setDTR(state)
  File "/usr/local/lib/python3.7/dist-packages/serial/serialutil.py", line 603, in setDTR
    self.dtr = value
  File "/usr/local/lib/python3.7/dist-packages/serial/serialutil.py", line 473, in dtr
    self._update_dtr_state()
  File "/usr/local/lib/python3.7/dist-packages/serial/serialposix.py", line 715, in _update_dtr_state
    fcntl.ioctl(self.fd, TIOCMBIC, TIOCM_DTR_str)
BrokenPipeError: [Errno 32] Broken pipe
*** [upload] Error 1
========================= [FAILED] Took 33.77 seconds

**Comment by gualo1983** [comment]
Riprogrammazione avvenuta con successo!
Avendo sostituito il cavo di connessione tra il nodemcu e il raspberry ho risolto il problema della porta di comunicazione tra i due.
Per la riprogrammazione ho risolto il problema uscendo da docker e installando direttamente esphome sul SO raspbian come da [guida](https://indomus.it/guide/configurare-compilare-e-installare-il-firmware-esphome-tramite-raspberry-pi-con-raspbian-o-linux/).

Una piccola precisazione se mi è permesso...
nella guida vi è scritto:

> INSTALLAZIONE SUL DISPOSITIVO TARGET
L’interfaccia Dashboard consente di installare il firmware solo in modalità OTA, cliccando su “UPLOAD” (il dispositivo deve essere acceso, collegato alla Wi-Fi e deve già possedere il firmware ESPHome – infatti questa procedura si usa per aggiornare);
Altrimenti (in caso di installazione ex-novo), si faccia sempre riferimento alle considerazioni alle modalità di riprogrammazione fatte sopra.

In realtà io son riuscito a far tutto dall'interfaccia grafica senza dover dare altri comandi nella shell del raspberry. 

Approfitto per ringraziare Snakuzzo e lo staff di inDomus

**Reply by xericcardo** [reply]
ora che riesco a vedere la porta quando provo ad inviare il bin mi da questo errore
Connecting........_
Chip is ESP8266EX
Features: WiFi
WARNING: Detected crystal freq 20.33MHz is quite different to normalized freq 26MHz. Unsupported crystal in use?
Crystal is 26MHz
MAC: 00:08:ff:00:00:00
Uploading stub...
Running stub...

A fatal error occurred: Invalid head of packet (0x46)
INFO Upload with baud rate 460800 failed. Trying again with baud rate 115200.
INFO Running:  esptool.py --before default_reset --after hard_reset --baud 115200 --chip esp8266 --port /dev/ttyUSB0 write_flash 0x0 /data/ble_tracker/.pioenvs/ble_tracker/firmware.bin
esptool.py v2.8
Serial port /dev/ttyUSB0
Connecting.....
Chip is ESP8266EX
Features: WiFi
WARNING: Detected crystal freq 20.45MHz is quite different to normalized freq 26MHz. Unsupported crystal in use?
Crystal is 26MHz
MAC: 00:08:ff:00:00:00
Uploading stub...
Running stub...

A fatal error occurred: Invalid head of packet (0x46)

**Reply by xericcardo** [reply]
risolto avevo sbagliato di impostare la board

**Reply by KC-inDomus** [reply]
> Una piccola precisazione se mi è permesso...
> nella guida vi è scritto:
> 
> > INSTALLAZIONE SUL DISPOSITIVO TARGET
> > L’interfaccia Dashboard consente di installare il firmware solo in modalità OTA, cliccando su “UPLOAD” (il dispositivo deve essere acceso, collegato alla Wi-Fi e deve già possedere il firmware ESPHome – infatti questa procedura si usa per aggiornare);
> > Altrimenti (in caso di installazione ex-novo), si faccia sempre riferimento alle considerazioni alle modalità di riprogrammazione fatte sopra.
> 
> In realtà io son riuscito a far tutto dall'interfaccia grafica senza dover dare altri comandi nella shell del raspberry.

si può fare in modi diversi. se prima fai solo compila, volendo poi puoi flashre cliccando successivamente su upload, oppure flashando contestualmente :)
